aries_vcx/src/handlers/issuance/credential_def/mod.rs

```diff
@@ -186,43 +147,99 @@ impl CredentialDef {
         )
     }
 
-    pub fn publish(mut self, tails_url: Option<&str>) -> VcxResult<Self> {
-        trace!("CredentialDef::publish >>>");
+    pub fn was_published(&self) -> bool {
+        self.state == PublicEntityStateType::Published
+    }
 
-        match _try_get_cred_def_from_ledger(&self.issuer_did, &self.cred_def_id) {
-            Ok(Some(ledger_cred_def_json)) => {
-                return Err(VcxError::from_msg(VcxErrorKind::CreateCredDef, format!("Credential definition with id {} already exists on the ledger: {}", self.cred_def_id, ledger_cred_def_json)));
+    pub fn publish_cred_def(self) -> VcxResult<Self> {
+        if let Some(ledger_cred_def_json) = _try_get_cred_def_from_ledger(&self.issuer_did, &self.cred_def_id)? {
+            return Err(VcxError::from_msg(VcxErrorKind::CredDefAlreadyCreated, format!("Credential definition with id {} already exists on the ledger: {}", self.cred_def_id, ledger_cred_def_json)));
+        }
+        anoncreds::publish_cred_def(&self.issuer_did, &self.cred_def_json)?;
+        Ok(
+            Self {
+                state: PublicEntityStateType::Published,
+                ..self
             }
-            Ok(None) => {
-                anoncreds::publish_cred_def(&self.issuer_did, &self.cred_def_json)?;
+        )
+    }
+
+    pub fn rotate_rev_reg(&mut self, revocation_details: RevocationDetails) -> VcxResult<RevocationRegistry> {
+        trace!("CredentialDef::rotate_rev_reg >>> revocation_details: {:?}", revocation_details);
+        let (tails_base_dir, max_creds) = (
+            revocation_details.clone().tails_base_dir.or(self.get_tails_base_dir()),
+            revocation_details.max_creds.or(self.get_max_creds())
+        );
 
-                match self.rev_reg {
-                    Some(ref mut rev_reg) => {
-                        if let Some(tails_url) = tails_url {
-                            rev_reg.rev_reg_def.value.tails_location = String::from(tails_url);
-                        } else {
-                            return Err(VcxError::from_msg(VcxErrorKind::InvalidOption, "tails_url must be specified if credential is revokable"));
-                        };
+        let new_rev_reg= match (&self.rev_reg, &tails_base_dir, &max_creds) {
+            (Some(rev_reg), Some(tails_base_dir), Some(max_creds)) => {
+                let tag_no =  rev_reg.tag + 1;
+                RevocationRegistry::create(&self.issuer_did, &self.cred_def_id, tails_base_dir, max_creds.clone(), tag_no)?
+            }
+            _ => return {
+                Err(VcxError::from_msg(VcxErrorKind::RevRegDefNotFound,
+                                       "No revocation registry definitions associated with this credential definition"
+                ))
+            }
+        };
+        self.rev_reg = Some(new_rev_reg.clone());
```

Comment:
This `clone()` would not be necessary without the log - I think it's worth removing it

Reply:
I had to clone because we both assigned new registry to `self.rev_reg` and also returned it from the function. I've modified the function so it would return `VcxResult<()>`.

Reply:
I verified the clone was not necessary before writing the comment.